.circleci/reportgenerator/main.go

```diff
@@ -0,0 +1,394 @@
+
+Link to failed build: ${linkToBuild}.
+
+Note: Information about any subsequent build failures that happen while
+this issue is open, will be added as comments with more information to this issue.
+`
+	issueCommentTemplate = `
+Link to latest failed build: ${linkToBuild}
+`
+)
+
+// getExistingIssues gathers an existing GitHub Issue related to previous failures
+// of the same job.
+func (rg *reportGenerator) getExistingIssue() error {
+	url := githubAPIBaseURL + "/search/issues"
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return fmt.Errorf("failed to prepare GET request: %v", err)
+	}
+
+	for k, v := range rg.headers {
+		req.Header.Set(k, v)
+	}
+
+	q := req.URL.Query()
+	q.Add("q",
+		fmt.Sprintf(
+			"%s in:title is:open repo:%s/%s",
+			rg.getIssueTitle(),
+			rg.envVariables[projectUsernameKey],
+			rg.envVariables[projectRepoNameKey],
+		),
+	)
+	req.URL.RawQuery = q.Encode()
+
+	rg.logger.Info("Searching GitHub for existing Issues")
+	resp, err := rg.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to get exiting GitHub issues: %v", err)
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read exiting GitHub issues: %v", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		rg.logger.Error(
+			"unexpected response from GitHub",
+			zap.String("status_code", string(resp.StatusCode)),
+			zap.String("response", string(body)),
+			zap.String("url", url),
+		)
+		return fmt.Errorf("unexpected response from GitHub")
+	}
+
+	var respUnmarshalled githubIssueSearchResult
+	if err = json.Unmarshal(body, &respUnmarshalled); err != nil {
+		return fmt.Errorf("failed to unmarshal response: %v", err)
+	}
+
+	switch respUnmarshalled.Count {
+	case 0:
+		// Nothing to do if there aren't any existing issues for the job.
+	case 1:
+		rg.issue = &respUnmarshalled.Issues[0]
+	default:
+		// If there are multiple issues matching the query, pick the one
+		// where there's an exact match in the Issue title. Unfortunately,
+		// GitHub API does not do an exact match.
+		rg.logger.Info(
+			"Received multiple issues. Looking for exact match",
+			zap.Int("num_issues", respUnmarshalled.Count),
+		)
+
+		requiredTitle := rg.getIssueTitle()
+		for _, issue := range respUnmarshalled.Issues {
+			if issue.Title == requiredTitle {
+				rg.issue = &issue
+				rg.logger.Info("Exact match found for Issue",
+					zap.String(
+						"html_url",
+						issue.HTMLURL,
+					),
+				)
+				return nil
+			}
+		}
+
+		rg.logger.Info("No exact match found for Issue",
+			zap.String(
+				"required_title",
+				requiredTitle,
+			),
+		)
+	}
+
+	return nil
+}
+
+// commentOnIssue adds a new comment on an existing GitHub issue with
+// information about the latest failure. This method is expected to be
+// called only if there's an existing open Issue for the current job.
+func (rg *reportGenerator) commentOnIssue() error {
```

Comment:
Maybe consider doing something like codecov bot does which it only ever posts 1 comment then when there's an update it amends the comment. It prevents generating a lot of extra e-mail/github notifications and from drowning out any real comments on the thread. There's pros and cons so I don't way is strictly better than the other.

Reply:
I think when we have issues per-test failure we should make it be updates. But I think for just job failures adding a new comment for now would be fine.

Reply:
I was planning on posting names of failed tests (wherever available) to the issue body for the first occurrence and then the same information as comments on subsequent build failures. It is rudimentary but would give us an idea of repeat offenders. 

We may also be able to use the CircleCI Insights API to make use of failure trends over a time.

Reply:
Oh, I was thinking there might be an issue per test failure (since the tests might be in different components and one person is responsible for one failure and another person responsible for a different failure). 

But I think it's probably good to start with the single one and can break it out later if needed.

For subsequent build failures I think it'd be nice to edit the existing issue instead of posting a new comment to limit the number of notifications generated. But we can go with the new comment first and see if it's really a problem.

Reply:
Here's an actual example of how it'll look: https://github.com/asuresh4/opentelemetry-collector/issues/29 with the current implementation.